### PR TITLE
Optimize NormalModuleFactory, UglifyJsPlugin, LoaderOptionsPlugin forEach to loop and more 

### PR DESCRIPTION
**Pull request by Tom910**
What kind of change does this PR introduce?
Refactoring

Did you add tests for your changes?
N/A

Summary
Rewrote frequently used forEach and no optimal structure

Does this PR introduce a breaking change?
No

### CODE REVIEW

**Comment by Tom910** [comment]
many-chunks
    HEAD 665ms ± 82ms [645ms; 684ms]
  √ should benchmark HEAD (1215234a2366a9f5a1b2054ac6579b6c39f8470e) (34719ms)
    v2.2.0 688ms ± 78ms [669ms; 707ms]
  √ should benchmark v2.2.0 (4be093d27942c25ec802a40268ed94e24d886b1d) (35328ms)
  √ HEAD should not be slower than v2.2.0 (4be093d27942c25ec802a40268ed94e24d886b1d)
    base 690ms ± 81ms [670ms; 710ms]
  √ should benchmark base (96ea95ad5e018bc0354900f7e98756a1ca3931d8) (35404ms)
  √ HEAD should not be slower than base (96ea95ad5e018bc0354900f7e98756a1ca3931d8)
many-modules
    HEAD 1005ms ± 145ms [963ms; 1047ms]
  √ should benchmark HEAD (1215234a2366a9f5a1b2054ac6579b6c39f8470e) (36453ms)
    v2.2.0 1040ms ± 116ms [1006ms; 1074ms]
  √ should benchmark v2.2.0 (4be093d27942c25ec802a40268ed94e24d886b1d) (36812ms)
  √ HEAD should not be slower than v2.2.0 (4be093d27942c25ec802a40268ed94e24d886b1d)
    base 1032ms ± 122ms [996ms; 1068ms]
  √ should benchmark base (96ea95ad5e018bc0354900f7e98756a1ca3931d8) (36260ms)
  √ HEAD should not be slower than base (96ea95ad5e018bc0354900f7e98756a1ca3931d8)
many-modules-source-map
    HEAD 1082ms ± 171ms [1031ms; 1134ms]
  √ should benchmark HEAD (1215234a2366a9f5a1b2054ac6579b6c39f8470e) (37403ms)
    v2.2.0 1102ms ± 148ms [1057ms; 1146ms]
  √ should benchmark v2.2.0 (4be093d27942c25ec802a40268ed94e24d886b1d) (37452ms)
  √ HEAD should not be slower than v2.2.0 (4be093d27942c25ec802a40268ed94e24d886b1d)
    base 1078ms ± 141ms [1036ms; 1120ms]
  √ should benchmark base (96ea95ad5e018bc0354900f7e98756a1ca3931d8) (37558ms)
  √ HEAD should not be slower than base (96ea95ad5e018bc0354900f7e98756a1ca3931d8)

**Comment by Tom910** [comment]
I watched the profile of our project and saw that a lot of time is spent optimize/UglifyJsPlugin.js

File: lib/optimize/RemoveParentModulesPlugin.js [hunk 1]
```diff
@@ -17,17 +17,21 @@ function chunkContainsModule(chunk, module) {
...
 
 function allHaveModule(someChunks, module, checkedChunks) {
 	if(!checkedChunks) checkedChunks = [];
 	var chunks = [];
 	for(var i = 0; i < someChunks.length; i++) {
-		checkedChunks.push(someChunks[i]);
-		var subChunks = hasModule(someChunks[i], module, checkedChunks);
+		var chunk = someChunks[i];
+
+		if(checkedChunks.indexOf(chunk) > 0) {
```
Comment (sokra) [inline lib/optimize/RemoveParentModulesPlugin.js]: Could we optimize this by making checkedChunks an `Map`?

File: lib/optimize/UglifyJsPlugin.js [hunk 2]
```diff
@@ -31,57 +73,51 @@ class UglifyJsPlugin {
...
 						}
 						uglify.base54.reset();
-						let ast = uglify.parse(input, {
+						var ast = uglify.parse(input, {
 							filename: file
 						});
 						if(options.compress !== false) {
 							ast.figure_out_scope();
-							let compress = uglify.Compressor(options.compress || {
+							var compress = uglify.Compressor(options.compress || {
```
Comment (sokra) [inline lib/optimize/UglifyJsPlugin.js]: Why are you changing all `let`s to `var`s?

Comment (Tom910) [inline lib/optimize/UglifyJsPlugin.js]: let and const, don't optimization in loop =( https://gist.github.com/chicoxyzzy/0ecb281f61120a4704d3313e4d600859#gistcomment-1975047

**Comment by Tom910** [comment]
> I don't think the changes in UglifyJsPlugin make sense.
> The number of files is not that big to optimize here..
I saw a significant cost for execution in UglifyJsPlugin on our project. And on the stage of additional asset processing on this and become optimized.

File: lib/optimize/RemoveParentModulesPlugin.js (hunk 1 above)
Comment (Tom910) [inline lib/optimize/RemoveParentModulesPlugin.js]: Ok, I'll check

**Comment by sokra** [comment]
> I saw a significant cost for execution in UglifyJsPlugin on our project. And on the stage of additional asset processing on this and become optimized.

ok, please add a benchmarkCase for this one.

**Comment by Tom910** [comment]
@sokra adding new benchmark case. Added by case with the conversion of files to prod mode

File: lib/optimize/RemoveParentModulesPlugin.js (hunk 1 above)
Comment (Tom910) [inline lib/optimize/RemoveParentModulesPlugin.js]: done

File: lib/optimize/UglifyJsPlugin.js (hunk 2 above)
Comment (wtgtybhertgeghgtwtg) [inline lib/optimize/UglifyJsPlugin.js]: Shouldn't a decision be made about that first?  There are a dozen PR's currently up moving `var` to `let` and `const`.

Comment (TheLarkInn) [inline lib/optimize/UglifyJsPlugin.js]: Yeah let's keep the let/const

Comment (Tom910) [inline lib/optimize/UglifyJsPlugin.js]: OK, i go back let

**Comment by Tom910** [comment]
The new result
```
    many-chunks
        HEAD 509ms ± 62ms [496ms; 522ms]
      ✓ should benchmark HEAD (d8fce987e5b2f1e5d666d7872a3520259ade943f) (33505ms)
        v2.2.0 578ms ± 69ms [563ms; 594ms]
      ✓ should benchmark v2.2.0 (4be093d27942c25ec802a40268ed94e24d886b1d) (33824ms)
======> HEAD is 14% faster than v2.2.0 (90% confidence)!
      ✓ HEAD should not be slower than v2.2.0 (4be093d27942c25ec802a40268ed94e24d886b1d)
```

**Comment by sokra** [comment]
`======> HEAD is 14% faster than v2.2.0 (90% confidence)!` this is not caused by this PR.

You need to compare against `base`.

**Comment by Tom910** [comment]
Result of travis

```
    many-chunks
        HEAD 826ms ± 105ms [798ms; 854ms]
      ✓ should benchmark HEAD (2d0ef88871d35cf9a1126ba1f40fd26f9f5bfe65) (34715ms)
        v2.2.0 889ms ± 104ms [861ms; 918ms]
      ✓ should benchmark v2.2.0 (4be093d27942c25ec802a40268ed94e24d886b1d) (35721ms)
======> HEAD is 8% faster than v2.2.0 (90% confidence)!
      ✓ HEAD should not be slower than v2.2.0 (4be093d27942c25ec802a40268ed94e24d886b1d)
        base 880ms ± 114ms [849ms; 911ms]
      ✓ should benchmark base (873aad11e8d6d5c466f5b5a60533f9a8822345f4) (35259ms)
      ✓ HEAD should not be slower than base (873aad11e8d6d5c466f5b5a60533f9a8822345f4)
```

File: lib/optimize/OccurrenceOrderPlugin.js [hunk 3]
```diff
@@ -16,36 +16,40 @@ class OccurrenceOrderPlugin {
...
 
 				function occurs(m) {
 					if(typeof m.__OccurenceOrderPlugin_occurs === "number") return m.__OccurenceOrderPlugin_occurs;
-					const result = m.reasons.map((r) => {
-						if(!r.module) return 0;
-						return r.module.chunks.length;
-					}).reduce((a, b) => {
-						return a + b;
-					}, 0) + m.chunks.length + m.chunks.filter((c) => {
-						c.entryModule === m;
```
Comment (Tom910) [inline lib/optimize/OccurrenceOrderPlugin.js]: @sokra There is likely a bug. If correct, then fall 4 test. I edit the code and tests?

**Comment by sokra** [comment]
This PR has grown to a state in which it's no longer mergable.

I'll only merge performance PR when 
* they either don't decrease readablility (forEach -> for do so)
* or they have a big impact on performance
* and they do very specific changes

rewriting all forEach to for is only acceptable if you make a single PR with a small changeset which has high effect on performance.

Other performance improvements can be accepted if they don't decrease readablility.

This PR is currently a gigant mix of everything which doesn't affect performance significant. Please split it into smaller PRs doing a change which can be proven to increase performance.

**Comment by Tom910** [comment]
@sokra 

> which it's no longer mergable.

This is only done in #4140 which was found in the PR and me)

> This PR is currently a gigant mix of everything which doesn't affect performance significant. Please split it into smaller PRs doing a change which can be proven to increase performance.
Ok, I break into many PR

**Pull request closed without merging**